---
title: "R Notebook"
output: html_notebook
---

#### Run startup function
```{r}
setwd("~/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/macro")
source("ui_init.R")
```

#### Obtain P4C metadata from Donovan et al. 2024:
# https://www.nature.com/articles/s41467-024-49781-1#Sec40
```{r}
# Donovan et al. 2024:
# https://www.nature.com/articles/s41467-024-49781-1#Sec40

fp.donovan2024_source <- "/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/rawdata/Source Data";
fn.donovan2024_source <- c("SourceData_Fig1.xlsx",
                           "SourceData_Fig2.xlsx",               
                           "SourceData_Fig3.xlsx",
                           "SourceData_Fig4.xlsx",
                           "SourceData_Fig5.xlsx",
                           "SourceData_Fig6.xlsx",
                           "SourceData_Fig7.xlsx",
                           "SourceData_Fig8.xlsx");
setwd(fp.donovan2024_source)
donovan2024_source_data <- fn.donovan2024_source %>%
  lapply(., read_excel_allsheets);

donovan2024_p4c_meta <- donovan2024_source_data %>%
  lapply(., rbindlist, fill = TRUE) %>%
  lapply(., select, c(SampleID, Karyotype, Age, Sex)) %>%
  lapply(., unique) %>%
  lapply(as.data.frame) %>%
  rbindlist(., fill = TRUE) %>%
  unique()

donovan2024_p4c_meta
```

#### Read P4C SomaLogic data downloaded from Synapse
```{r}
p4c_soma <- fread("/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/rawdata/HTP_SOMAscan_Proteomics_Synapse.txt",
                  sep = "\t")

p4c_soma
```

#### Assess the overlap of the Somalogic publication LabIDs and the Donovan 2024 Sample IDs
```{r}
p4c_soma %>%
  select(LabID) %>%
  unique() %>%
  mutate(In_proteomics_paper = 1) %>%
  full_join(donovan2024_p4c_meta %>%
              mutate(In_donovan2024 = 1),
            by = c("LabID" = "SampleID")) %>%
  group_by(In_proteomics_paper, In_donovan2024) %>%
  summarise(N = n())

keep_labids <- (p4c_soma %>%
  select(LabID) %>%
  unique() %>%
  mutate(In_proteomics_paper = 1) %>%
  full_join(donovan2024_p4c_meta %>%
              mutate(In_donovan2024 = 1),
            by = c("LabID" = "SampleID")) %>%
  filter(!is.na(In_proteomics_paper) & !is.na(In_donovan2024)) %>%
  select(LabID) %>%
  unique())$LabID
```

#### Prepare published SomaLogic data and add metadata
```{r}
# https://www.nature.com/articles/s41591-019-0673-2
# https://pmc.ncbi.nlm.nih.gov/articles/PMC7062043/#SD2
st7 <- openxlsx::read.xlsx("/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/rawdata/NIHMS1541641-supplement-1541641_Sup_Tab1-17.xlsx",
                 sheet = 8, # "ST7 Plasma proteomic clock"
                 startRow = 2) %>% 
  `rownames<-`(NULL)

#install.packages("openxlsx")
#library(openxlsx)
st11_clusterIDs <- openxlsx::read.xlsx("/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/rawdata/NIHMS1541641-supplement-1541641_Sup_Tab1-17.xlsx",
                 sheet = 12, # "ST7 Plasma proteomic clock"
                 startRow = 3) %>% 
  `rownames<-`(NULL)

proteomic_clock_proteins <- st7 %>% select(Variable) %>% filter(Variable != "(Intercept)")

# st7
#st11_clusterIDs
#proteomic_clock_proteins

proteomic_clock_proteins_clusters <- proteomic_clock_proteins %>%
  full_join(st11_clusterIDs, by = c("Variable" = "Protein")) %>%
  separate(Variable, into = paste0("v", seq(1:10)), remove = FALSE, extra = "merge") %>%
  gather(key = "key", value = "value", v1:v10) %>%
  mutate(value = case_when(is.na(value) ~ "", .default = value),
         key = gsub("v", "", key) %>% as.numeric()) %>%
  group_by(Variable, ClusterID) %>%
  arrange(Variable, key) %>%
  filter(value != "") %>%
  mutate(max_key = max(key)) %>%
  mutate(is_last_piece = case_when(key == max_key ~ 1, .default = 0)) %>%
  filter(is_last_piece == 0) %>%
  select(-c(max_key, is_last_piece)) %>%
  mutate(key = paste0("v", key)) %>%
  summarise(Aptamer = paste(value, collapse = "."));

rm(st11_clusterIDs, proteomic_clock_proteins); gc()

st7
proteomic_clock_proteins_clusters
```

#### Evaluate the overlap of proteins in P4C's SomaLogic vs. in the published proteomic clock proteins
```{r}
a <- p4c_soma %>% select(Aptamer) %>% unique() %>% mutate(In_P4C = 1)
b <- proteomic_clock_proteins_clusters %>% mutate(In_st7 = 1)

a %>%
  full_join(b,  by = c("Aptamer" = "Aptamer")) %>%
  mutate(In_P4C = case_when(is.na(In_P4C) ~ 0, .default = In_P4C),
         In_st7 = case_when(is.na(In_st7) ~ 0, .default = In_st7),
         In_sum = In_P4C + In_st7) %>%
  group_by(In_P4C, In_st7, In_sum) %>%
  summarise(N = n())
```

```{r}
# p4c_soma %>% select(Aptamer) %>% unique() %>%
#   mutate(In_published_p4c_soma = 1) %>%
#   full_join(proteomic_clock_proteins_clusters %>% mutate(In_st17 = 1), by = "Aptamer") %>%
#   group_by(In_published_p4c_soma, In_st17) %>%
#   summarise(N = n())

keep_aptamers <- p4c_soma %>%
  select(Aptamer) %>%
  unique() %>%
  mutate(In_published_p4c_soma = 1) %>%
  full_join(proteomic_clock_proteins_clusters %>% mutate(In_st17 = 1), by = "Aptamer") %>%
  filter(!is.na(In_published_p4c_soma) & !is.na(In_st17)) %>%
  select(Aptamer, Variable)

keep_aptamers
```

```{r}
gene_chr_anno <- fread("/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/rawdata/martquery_0127162849_293.txt.gz") %>%
  select(`UniProtKB/Swiss-Prot ID`, `Gene name`, `Gene Synonym`,
         `Chromosome/scaffold name`) %>%
  rename(biomart_gene_name = `Gene name`,
         biomart_gene_synonym = `Gene Synonym`) %>%
  unique() %>%
  arrange(`UniProtKB/Swiss-Prot ID`) %>%
  filter(!is.na(`UniProtKB/Swiss-Prot ID`) & `UniProtKB/Swiss-Prot ID`!="") %>%
  mutate(In_biomart_anno = 1) %>%
  rename(UniProt = `UniProtKB/Swiss-Prot ID`) %>%
  #select(`Chromosome/scaffold name`) %>%
  #unique() %>%
  mutate(chromosome = case_when( !is.na(as.numeric(`Chromosome/scaffold name`)) ~
                                   as.numeric(`Chromosome/scaffold name`),
                                 grepl("CHR10", `Chromosome/scaffold name`)==TRUE ~ 10,
                                 grepl("CHR11", `Chromosome/scaffold name`)==TRUE ~ 11,
                                 grepl("CHR12", `Chromosome/scaffold name`)==TRUE ~ 12,
                                 grepl("CHR13", `Chromosome/scaffold name`)==TRUE ~ 13,
                                 grepl("CHR14", `Chromosome/scaffold name`)==TRUE ~ 14,
                                 grepl("CHR15", `Chromosome/scaffold name`)==TRUE ~ 15,
                                 grepl("CHR16", `Chromosome/scaffold name`)==TRUE ~ 16,
                                 grepl("CHR17", `Chromosome/scaffold name`)==TRUE ~ 17,
                                 grepl("CHR18", `Chromosome/scaffold name`)==TRUE ~ 18,
                                 grepl("CHR19", `Chromosome/scaffold name`)==TRUE ~ 19,
                                 grepl("CHR20", `Chromosome/scaffold name`)==TRUE ~ 20,
                                 grepl("CHR21", `Chromosome/scaffold name`)==TRUE ~ 21,
                                 grepl("CHR22", `Chromosome/scaffold name`)==TRUE ~ 22,
                                 grepl("CHR23", `Chromosome/scaffold name`)==TRUE ~ 23,
                                 
                                 grepl("CHR1", `Chromosome/scaffold name`)==TRUE ~ 1,
                                 grepl("CHR2", `Chromosome/scaffold name`)==TRUE ~ 2,
                                 grepl("CHR3", `Chromosome/scaffold name`)==TRUE ~ 3,
                                 grepl("CHR4", `Chromosome/scaffold name`)==TRUE ~ 4,
                                 grepl("CHR5", `Chromosome/scaffold name`)==TRUE ~ 5,
                                 grepl("CHR6", `Chromosome/scaffold name`)==TRUE ~ 6,
                                 grepl("CHR7", `Chromosome/scaffold name`)==TRUE ~ 7,
                                 grepl("CHR8", `Chromosome/scaffold name`)==TRUE ~ 8,
                                 grepl("CHR9", `Chromosome/scaffold name`)==TRUE ~ 9,
                                 
                                 grepl("CHRX", `Chromosome/scaffold name`)==TRUE ~ 22,
                                 grepl("CHRY", `Chromosome/scaffold name`)==TRUE ~ 23,
                                 
                                 grepl("MT", `Chromosome/scaffold name`)==TRUE ~ 24, # Use chromosome=24 for mitochondrial DNA
                                 
                                 `Chromosome/scaffold name`=="X" ~ 22,
                                 `Chromosome/scaffold name`=="Y" ~ 23,
                                 `Chromosome/scaffold name`=="MT" ~ 24,

                                 .default = NA)) %>%
  arrange(`Chromosome/scaffold name`) %>%
  rename(biomart_chr_scaffold = `Chromosome/scaffold name`)

gene_chr_anno
```

#### ai1_p4c_soma
```{r}
ai_p4c_soma <-  p4c_soma %>%
  mutate(In_published_p4c_soma = 1) %>%
  full_join(proteomic_clock_proteins_clusters %>%
              mutate(In_st17 = 1),
            by = "Aptamer") %>%
  full_join(donovan2024_p4c_meta %>% rename(LabID = SampleID),
            by = c("LabID")) %>%
  left_join(gene_chr_anno, by = "UniProt") %>%
  filter(LabID %in% keep_labids) %>%
  rename(Abundance = Value) %>%
  mutate(log10_Abundance = log10(Abundance))

ai_p4c_soma
```

```{r}
ai_p4c_soma_d21_sstats <- p4c_soma %>%
  mutate(In_published_p4c_soma = 1) %>%
  full_join(proteomic_clock_proteins_clusters %>%
              mutate(In_st17 = 1),
            by = "Aptamer") %>%
  full_join(donovan2024_p4c_meta %>% rename(LabID = SampleID),
            by = c("LabID")) %>%
  left_join(gene_chr_anno, by = "UniProt") %>%
  filter(LabID %in% keep_labids) %>%
  rename(Abundance = Value) %>%
  mutate(log10_Abundance = log10(Abundance)) %>%
  filter(Karyotype == "Control") %>%
  group_by(Aptamer, Karyotype) %>%
  summarise(mean_log10Abundance.D21 = mean(log10_Abundance, na.rm = TRUE),
            sd_log10Abundance.D21 = sd(log10_Abundance, na.rm = TRUE)) %>%
  ungroup();

ai_p4c_soma_d21_sstats

setwd("~/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/aidata")
fwrite(ai_p4c_soma_d21_sstats, "ai_p4c_soma_d21_sstats.csv.gz")
```

#### ai_p4c_soma_t21_sstats
```{r}
ai_p4c_soma_t21_sstats <- p4c_soma %>%
  mutate(In_published_p4c_soma = 1) %>%
  full_join(proteomic_clock_proteins_clusters %>%
              mutate(In_st17 = 1),
            by = "Aptamer") %>%
  full_join(donovan2024_p4c_meta %>% rename(LabID = SampleID),
            by = c("LabID")) %>%
  left_join(gene_chr_anno, by = "UniProt") %>%
  filter(LabID %in% keep_labids) %>%
  rename(Abundance = Value) %>%
  mutate(log10_Abundance = log10(Abundance)) %>%
  filter(Karyotype == "T21") %>%
  group_by(Aptamer, Karyotype) %>%
  summarise(mean_log10Abundance.T21 = mean(log10_Abundance, na.rm = TRUE),
            sd_log10Abundance.T21 = sd(log10_Abundance, na.rm = TRUE)) %>%
  ungroup()

ai_p4c_soma_t21_sstats

setwd("~/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/aidata")
fwrite(ai_p4c_soma_t21_sstats, "ai_p4c_soma_t21_sstats.csv.gz")
```

#### ar_p4c_soma
```{r}
# ar_p4c_soma = ai1_p4c_soma + ai_p4c_soma_d21_sstats + ai_p4c_soma_t21_sstats

ar_p4c_soma <- ai_p4c_soma %>%
  full_join(ai_p4c_soma_d21_sstats %>% select(Aptamer, mean_log10Abundance.D21, sd_log10Abundance.D21),
            by = c("Aptamer")) %>%
  full_join(ai_p4c_soma_t21_sstats %>% select(Aptamer, mean_log10Abundance.T21, sd_log10Abundance.T21),
            by = "Aptamer") %>%
  group_by(Aptamer) %>%
  mutate(Z_log10_Abundance = (log10_Abundance - mean_log10Abundance.D21)/sd_log10Abundance.D21) %>%
  ungroup() %>%
  rename(Cluster = ClusterID) %>%
  mutate(Cluster = paste0("Cluster ", Cluster),
         Cluster = factor(Cluster,
                          levels = c("Cluster 7",
                                     "Cluster 6", 
                                     "Cluster 3",
                                     "Cluster 4", 
                                     "Cluster 2", 
                                     "Cluster 1",
                                     "Cluster 5",
                                     "Cluster 8"))) %>%
  group_by(Cluster) %>%
  mutate(ggtitle = Cluster) %>%
  ungroup() %>%
  arrange(Cluster) %>%
  mutate(T21 = case_when(Karyotype == "T21" ~ 1,
                         Karyotype == "Control" ~ 0,
                         .default = NA),
         Female = case_when(Sex == "Female" ~ 1,
                            Sex == "Male" ~ 0,
                            .default = NA)) %>%
  mutate(Cluster = case_when(is.na(Cluster) ~ "Cluster Other", .default = Cluster)) %>%
  mutate(Aptamer_Chr_Cluster = paste0(Aptamer, " (Chr", chromosome, ", ", Cluster, ")"),
         UniProt_Aptamer_Chr_Cluster = paste0(UniProt, "|", Aptamer, " (Chr", chromosome, ", ", Cluster, ")")) %>%
  select(UniProt_Aptamer_Chr_Cluster, Aptamer_Chr_Cluster, everything())

ar_p4c_soma
ar_p4c_soma %>% select(Cluster) %>% unique()

setwd("~/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/ardata")
fwrite(ar_p4c_soma, "ar_p4c_soma.csv.gz")
```
```{r}
ar_p4c_soma.split <- ar_p4c_soma %>% split(., .$Cluster)
```

```{r}
gg_seT <- list()
for ( i in 1:length(ar_p4c_soma.split) ){
  gg_seT[[i]] <- ar_p4c_soma.split[[i]] %>%
    ggplot(aes(x = Age, y = Z_log10_Abundance, color = Karyotype)) +
    scale_colour_manual(values = GrayBlue) +
    geom_smooth(se = TRUE) +
    theme_bw() +
    ggtitle(ggData.split[[i]]$ggtitle[1]) +
    coord_cartesian(xlim = c(0,100),
                    ylim = c(-2,2)) +
    scale_x_continuous("Age (years)", 
                       labels = as.character(seq(0, 100, by=20)),
                       breaks = seq(0, 100, by = 20), 
                       expand=c(0,0)
                       ) +
    scale_y_continuous("Protein levels (Z-score)", 
                       labels = as.character(seq(-2, 2, by=1)),
                       breaks = seq(-2, 2, by = 1), 
                       expand=c(0,0)
                       ) +
    theme(aspect.ratio = 1.0,
          legend.position = "bottom")
}

#gg_seT[[8]]

for ( i in 1:length(ar_p4c_soma.split) ) {
  setwd("/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/output")
  ggsave(gg_seT[[i]],
         filename = paste("20250124_Proteomic_Aging_", ar_p4c_soma.split[[i]]$ggtitle[1], "_seTRUE_v0.2.png", sep = ""),
         width = 5, height = 5, units = "in")
}

gg_seT
```

#### ar_p4c_soma_sumZcluster
```{r}
ar_p4c_soma_sumZcluster <- ai_p4c_soma %>%
  full_join(ai_p4c_soma_d21_sstats %>% select(Aptamer, mean_log10Abundance.D21, sd_log10Abundance.D21),
            by = c("Aptamer")) %>%
  full_join(ai_p4c_soma_t21_sstats %>% select(Aptamer, mean_log10Abundance.T21, sd_log10Abundance.T21),
            by = "Aptamer") %>%
  group_by(Aptamer) %>%
  mutate(Z_log10_Abundance = (log10_Abundance - mean_log10Abundance.D21)/sd_log10Abundance.D21) %>%
  ungroup() %>%
  rename(Cluster = ClusterID) %>%
  mutate(Cluster = paste0("Cluster ", Cluster),
         Cluster = factor(Cluster,
                          levels = c("Cluster 7",
                                     "Cluster 6", 
                                     "Cluster 3",
                                     "Cluster 4", 
                                     "Cluster 2", 
                                     "Cluster 1",
                                     "Cluster 5",
                                     "Cluster 8"))) %>%
  group_by(Cluster) %>%
  mutate(ggtitle = Cluster) %>%
  ungroup() %>%
  arrange(Cluster) %>%
  mutate(T21 = case_when(Karyotype == "T21" ~ 1,
                         Karyotype == "Control" ~ 0,
                         .default = NA),
         Female = case_when(Sex == "Female" ~ 1,
                            Sex == "Male" ~ 0,
                            .default = NA)) %>%
  mutate(Cluster = case_when(is.na(Cluster) ~ "Cluster Other", .default = Cluster)) %>%
  group_by(LabID,
           T21, Karyotype, Female, Sex, Age,
           Cluster) %>%
  summarise(sumZ_Cluster = sum(Z_log10_Abundance)) %>%
  ungroup() %>%
  mutate(Cluster = gsub(" ", "_", Cluster)) %>%
  spread(key = Cluster, value = sumZ_Cluster) %>%
  gather(key = "Cluster", value = "sumZ_Cluster", Cluster_1:Cluster_Other)

ar_p4c_soma_sumZcluster

ar_p4c_soma_sumZcluster.split <- ar_p4c_soma_sumZcluster %>%
  split(., .$Cluster)

# IMPORTANT NOTE: If NA for any cluster, exclude participant from aging Z score analysis and possibly from all analyses (decide later...)

setwd("~/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/ardata")
fwrite(ar_p4c_soma_sumZcluster, "ar_p4c_soma_sumZcluster.csv.gz")
```

#### ar_p4c_soma_sumZcluster_agePredT21
```{r}
ar_p4c_soma_sumZcluster_agePredT21 <- ai_p4c_soma %>%
  full_join(ai_p4c_soma_d21_sstats %>% select(Aptamer, mean_log10Abundance.D21, sd_log10Abundance.D21),
            by = c("Aptamer")) %>%
  full_join(ai_p4c_soma_t21_sstats %>% select(Aptamer, mean_log10Abundance.T21, sd_log10Abundance.T21),
            by = "Aptamer") %>%
  group_by(Aptamer) %>%
  mutate(Z_log10_Abundance = (log10_Abundance - mean_log10Abundance.D21)/sd_log10Abundance.D21) %>%
  ungroup() %>%
  rename(Cluster = ClusterID) %>%
  mutate(Cluster = paste0("Cluster ", Cluster),
         Cluster = factor(Cluster,
                          levels = c("Cluster 7",
                                     "Cluster 6", 
                                     "Cluster 3",
                                     "Cluster 4", 
                                     "Cluster 2", 
                                     "Cluster 1",
                                     "Cluster 5",
                                     "Cluster 8"))) %>%
  group_by(Cluster) %>%
  mutate(ggtitle = Cluster) %>%
  ungroup() %>%
  arrange(Cluster) %>%
  mutate(T21 = case_when(Karyotype == "T21" ~ 1,
                         Karyotype == "Control" ~ 0,
                         .default = NA),
         Female = case_when(Sex == "Female" ~ 1,
                            Sex == "Male" ~ 0,
                            .default = NA)) %>%
  mutate(Cluster = case_when(is.na(Cluster) ~ "Cluster Other", .default = Cluster)) %>%
  group_by(LabID,
           T21, Karyotype, Female, Sex, Age,
           Cluster) %>%
  summarise(sumZ_Cluster = sum(Z_log10_Abundance)) %>%
  ungroup() %>%
  mutate(Cluster = gsub(" ", "_", Cluster)) %>%
  spread(key = Cluster, value = sumZ_Cluster) %>%
  gather(key = "Cluster", value = "sumZ_Cluster", Cluster_1:Cluster_Other) %>%
  mutate(Cluster = paste0("sumZ_", Cluster)) %>%
  filter(T21 == 1) %>%
  spread(key = Cluster, value = sumZ_Cluster)

ar_p4c_soma_sumZcluster_agePredT21

setwd("~/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/ardata")
fwrite(ar_p4c_soma_sumZcluster_agePredT21,
       "ar_p4c_soma_sumZcluster_agePredT21.csv.gz")
```

#### Linear model for log(Age) vs. sumZ_Cluster_1 + ... + sumZ_Cluster8
```{r}
set.seed(1234)
fit.sumZcluster_agePredT21 <- lm(log(Age) ~ sumZ_Cluster_1 +
                                sumZ_Cluster_2+
                                sumZ_Cluster_3 +
                                sumZ_Cluster_4 +
                                sumZ_Cluster_5 +
                                sumZ_Cluster_6 +
                                  sumZ_Cluster_7 +
                                  sumZ_Cluster_8 +
                                  #sumZ_Cluster_Other +
                                Female
                                ,
                                data = ar_p4c_soma_sumZcluster_agePredT21);

glance.sumZcluster_agePredT21 <- fit.sumZcluster_agePredT21 %>% glance()
glance.sumZcluster_agePredT21

tidy.sumZcluster_agePredT21 <- fit.sumZcluster_agePredT21 %>%
  tidy(conf.int = TRUE) %>%
    filter(term!="(Intercept)") %>%
  rename(log10FC = estimate,
         P = p.value) %>%
  select(-c(std.error, statistic)) %>%
  arrange(P);
tidy.sumZcluster_agePredT21

set.seed(1234)
fit.sumZcluster_agePredT21_MIN <- lm(log(Age) ~ #sumZ_Cluster_1 +
                                #sumZ_Cluster_2+
                                #sumZ_Cluster_3 +
                                sumZ_Cluster_4 +
                                sumZ_Cluster_5 +
                                #sumZ_Cluster_6 +
                                  #sumZ_Cluster_7 +
                                  sumZ_Cluster_8 +
                                  #sumZ_Cluster_Other +
                                Female
                                ,
                                data = ar_p4c_soma_sumZcluster_agePredT21);

tidy.sumZcluster_agePredT21_MIN <- fit.sumZcluster_agePredT21_MIN %>%
  tidy(conf.int = TRUE) %>%
    filter(term!="(Intercept)") %>%
  rename(log10FC = estimate,
         P = p.value) %>%
  select(-c(std.error, statistic)) %>%
  arrange(P);
tidy.sumZcluster_agePredT21_MIN

tidy.sumZcluster_agePredT21 <- fit.sumZcluster_agePredT21 %>%
  tidy(conf.int = TRUE) %>%
    filter(term!="(Intercept)") %>%
  rename(log10FC = estimate,
         P = p.value) %>%
  select(-c(std.error, statistic)) %>%
  arrange(P);

glance.sumZcluster_agePredT21 <- fit.sumZcluster_agePredT21_MIN %>% glance()
glance.sumZcluster_agePredT21
```

#### lm(sumZ_Cluster ~ T21 + Age + Female, data = .)
```{r}
set.seed(1234)
fit.sumZcluster_modelA <- ar_p4c_soma_sumZcluster %>%
  split(., .$Cluster) %>%
  map(~lm(sumZ_Cluster ~ T21 + Age + Female, data = .));

dd_tidy_sumZcluster_modelA <-  ar_p4c_soma_sumZcluster %>%
  split(., .$Cluster) %>%
  map2(.x = fit.sumZcluster_modelA, .y = .,
       .f = ~tidy(x = .x, conf.int=TRUE, data = .y)) %>%
    bind_rows(.id="Cluster") %>%
    filter(term!="(Intercept)") %>%
  mutate(Padj_BH = p.adjust(p.value, method = "BH")) %>%
  ungroup();

dd_tidy_sumZcluster_modelA %>%
  filter(term == "T21") %>%
  arrange(p.value) %>%
  select(c(std.error, statistic));

setwd("~/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/ddata")
fwrite(dd_tidy_sumZcluster_modelA, "dd_tidy_sumZcluster_modelA.csv")
```

#### #### lm(sumZ_Cluster ~ T21 + Age + Female + T21*Age, data = .)
```{r}
set.seed(1234)
fit.sumZcluster_modelB <-  ar_p4c_soma_sumZcluster %>%
  split(., .$Cluster) %>%
    map(~lm(sumZ_Cluster ~ T21 + Age + Female + T21*Age, data = .));

dd_tidy_sumZcluster_modelB <-  ar_p4c_soma_sumZcluster %>%
  split(., .$Cluster) %>%
  map2(.x = fit.sumZcluster_modelB, .y = .,
       .f = ~tidy(x = .x, conf.int=TRUE, data = .y)) %>%
    bind_rows(.id="Cluster") %>%
    filter(term!="(Intercept)") %>%
  group_by(term) %>%
  mutate(Padj_BH = p.adjust(p.value, method = "BH")) %>%
  ungroup();

dd_tidy_sumZcluster_modelB %>%
  filter(term == "T21") %>%
  arrange(p.value) %>%
  select(-c(std.error, statistic));
dd_tidy_sumZcluster_modelB %>%
  filter(term == "T21:Age") %>%
  arrange(p.value) %>%
  select(-c(std.error, statistic));

dd_tidy_sumZcluster_modelB

setwd("~/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/ddata")
fwrite(dd_tidy_sumZcluster_modelB, "dd_tidy_sumZcluster_modelB.csv")
```

#### dd_tidy_Zlog10Abundance_modelA
```{r}
set.seed(1234)
fit.Zlog10Abundance_modelA <- ar_p4c_soma %>%
  split(., .$UniProt_Aptamer_Chr_Cluster) %>%
  map(~lm(Z_log10_Abundance ~ T21 + Age + Female, data = .));

dd_tidy_Zlog10Abundance_modelA <- ar_p4c_soma %>%
  split(., .$UniProt_Aptamer_Chr_Cluster) %>%
  map2(.x = fit.Zlog10Abundance_modelA, .y = .,
       .f = ~tidy(x = .x, conf.int=TRUE, data = .y)) %>%
    bind_rows(.id="UniProt_Aptamer_Chr_Cluster") %>%
    filter(term!="(Intercept)") %>%
  mutate(Padj_BH = p.adjust(p.value, method = "BH")) %>%
  ungroup() %>%
  rename(Zlog10_FC = estimate,
         P = p.value) %>%
  mutate(cat_Zlog10_FC = case_when(Zlog10_FC < 0 ~ "Down",
                                   Zlog10_FC > 0 ~ "Up",
                                   .default = NA)) %>%
  separate(UniProt_Aptamer_Chr_Cluster, sep = "[|]", extra = "merge", remove = FALSE,
           into = c("UniProt", "b1")) %>%
  select(-c(b1)) %>%
  separate(UniProt_Aptamer_Chr_Cluster, sep = "[(]", extra = "merge", remove = FALSE,
           into = c("a", "b")) %>%
  select(-c(a)) %>%
  separate(b, sep = ",", extra = "merge", remove = FALSE,
           into = c("chromosome", "b2")) %>%
  select(-c(b, b2)) %>%
  mutate(Sig10_PadjBH = case_when(Padj_BH<0.10 ~ 1, .default = 0),
         Sig05_PadjBH = case_when(Padj_BH<0.05 ~ 1, .default = 0),
         cat_Zlog10_FC = case_when(Zlog10_FC < 0 ~ "-",
                                   Zlog10_FC > 0 ~ "+",
                                   .default = NA));

dd_tidy_Zlog10Abundance_modelA

setwd("~/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/ddata")
fwrite(dd_tidy_Zlog10Abundance_modelA, "dd_tidy_Zlog10Abundance_modelA.csv.gz")
```

```{r}
set.seed(1234)
fit.Zlog10Abundance_modelB <- ar_p4c_soma %>%
  split(., .$UniProt_Aptamer_Chr_Cluster) %>%
  map(~lm(Z_log10_Abundance ~ T21 + Age + Female + T21*Age, data = .));

dd_tidy_Zlog10Abundance_modelB <- ar_p4c_soma %>%
  head() %>%
  split(., .$UniProt_Aptamer_Chr_Cluster) %>%
  map2(.x = fit.Zlog10Abundance_modelB, .y = .,
       .f = ~tidy(x = .x, conf.int=TRUE, data = .y)) %>%
    bind_rows(.id="UniProt_Aptamer_Chr_Cluster") %>%
    filter(term!="(Intercept)") %>%
  mutate(Padj_BH = p.adjust(p.value, method = "BH")) %>%
  ungroup() %>%
  rename(Zlog10_FC = estimate,
         P = p.value) %>%
  mutate(cat_Zlog10_FC = case_when(Zlog10_FC < 0 ~ "Down",
                                   Zlog10_FC > 0 ~ "Up",
                                   .default = NA)) %>%
  separate(UniProt_Aptamer_Chr_Cluster, sep = "[|]", extra = "merge", remove = FALSE,
           into = c("UniProt", "b1")) %>%
  select(-c(b1)) %>%
  separate(UniProt_Aptamer_Chr_Cluster, sep = "[(]", extra = "merge", remove = FALSE,
           into = c("a", "b")) %>%
  select(-c(a)) %>%
  separate(b, sep = ",", extra = "merge", remove = FALSE,
           into = c("chromosome", "b2")) %>%
  select(-c(b, b2)) %>%
  mutate(Sig10_PadjBH = case_when(Padj_BH<0.10 ~ 1, .default = 0),
         Sig05_PadjBH = case_when(Padj_BH<0.05 ~ 1, .default = 0),
         cat_Zlog10_FC = case_when(Zlog10_FC < 0 ~ "Down in T21",
                                   Zlog10_FC > 0 ~ "Up in T21",
                                   .default = NA));

dd_tidy_Zlog10Abundance_modelB

setwd("~/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/ddata")
fwrite(dd_tidy_Zlog10Abundance_modelB, "dd_tidy_Zlog10Abundance_modelB.csv.gz")
```

```{r}
dd_tidy_Zlog10Abundance_modelA %>%
  filter(term == "T21") %>%
  arrange(P) %>%
  select(-c(std.error, statistic)) %>%
  select(Aptamer_Chr_Cluster, everything());

dd_tidy_Zlog10Abundance_modelB %>%
  filter(term == "T21") %>%
  arrange(P) %>%
  select(-c(std.error, statistic)) %>%
  select(Aptamer_Chr_Cluster, everything());
tidy.Zlog10Abundance_modelB %>%
  filter(term == "T21:Age") %>%
  arrange(P) %>%
  select(-c(std.error, statistic)) %>%
  select(Aptamer_Chr_Cluster, everything());

dd_tidy_Zlog10Abundance_modelB %>%
  filter(term == "T21:Age") %>%
  arrange(P) %>%
  select(-c(std.error, statistic)) %>%
  select(Aptamer_Chr_Cluster, everything()) %>%
  mutate(Sig10_PadjBH = case_when(Padj_BH<0.10 ~ 1, .default = 0),
         Sig05_PadjBH = case_when(Padj_BH<0.05 ~ 1, .default = 0)) %>%
  filter(grepl("Chr21", Aptamer_Chr_Cluster)==TRUE) %>%
  filter(Sig10_PadjBH == 1);
```

#### dd_ggbar_chr_sigT21_modelA
```{r}
dd_ggbar_chr_sigT21_modelA <- dd_tidy_Zlog10Abundance_modelA %>%
  arrange(P) %>%
  select(-c(std.error, statistic)) %>%
  select(Aptamer_Chr_Cluster, everything()) %>%
  mutate(SigPadjBH10 = case_when(Padj_BH<0.10 ~ "SigPadjBH10_Yes", .default = "SigPadjBH10_No"),
         SigPadjBH05 = case_when(Padj_BH<0.05 ~ "SigPadjBH05_Yes", .default = "SigPadjBH05_No")) %>%
  filter(term %in% c("T21")) %>%
  group_by(chromosome, term, SigPadjBH10) %>%
  summarise(N_Aptamers_SigPadjBH10_T21byAge = n()) %>%
  ungroup() %>%
  spread(key = SigPadjBH10, value = N_Aptamers_SigPadjBH10_T21byAge) %>%
  mutate(N_Total_Aptamers_Chr = SigPadjBH10_No + SigPadjBH10_Yes,
         Pct_SigPadjBH.char = paste0(as.character(round(100*SigPadjBH10_Yes/N_Total_Aptamers_Chr)),
                                "%"),
         Pct_SigPadjBH = 100*SigPadjBH10_Yes/N_Total_Aptamers_Chr,
         # gg_chromosome = factor(chromosome, levels = c(paste0("Chr", chromosome))),
         gg_chromosome = as.numeric(gsub("Chr", "", chromosome)),
         gg_chromosome = case_when(is.na(gg_chromosome) ~ -1, .default = gg_chromosome)) %>%
  arrange(gg_chromosome) %>%
  mutate(chr21cat = case_when(chromosome == "Chr21" ~ "Chromosome 21",
                              .default = "Other chromosome"));

dd_ggbar_chr_sigT21_modelA

setwd("/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/ddata")
fwrite(dd_ggbar_chr_sigT21_modelA, "dd_ggbar_chr_sigT21_modelA.csv.gz")
```

#### f_ggbar_chr_sigT21_modelA
```{r}
f_ggbar_chr_sigT21_modelA <- dd_ggbar_chr_sigT21_modelA %>%
  filter(term == "T21") %>%
  ggplot(aes(x = gg_chromosome, y = Pct_SigPadjBH, color = chr21cat, fill = chr21cat)) +
  geom_bar(position = "dodge", stat = "identity") +
  theme_bw() +
  theme(legend.position = "none",
        aspect.ratio = 1.0) +
  xlab("Chromosome") +
  ylab("Aptamers (%)") +
  ggtitle("Chromosomal distribution of aptamers with\nsignificantly differential abundance in T21") +
  scale_x_continuous("Chromosome", 
                       labels = as.character(seq(-1, 24, by=1)),
                       breaks = seq(-1, 24, by = 1), 
                       expand=c(0,0))

setwd("/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/output")
ggsave(f_ggbar_chr_sigT21_modelA,
         filename = "f_ggbar_chr_sigT21_modelA.png",
         width = 5, height = 5, units = "in")
```

```{r}
# dd_ggbar_chr_sigT21_modelA
ggData_chr_distribution_SigPadjBH10.modelB02 <- tidy.Zlog10Abundance_modelB02 %>%
  arrange(P) %>%
  select(-c(std.error, statistic)) %>%
  select(Aptamer_Chr_Cluster, everything()) %>%
  mutate(SigPadjBH10 = case_when(Padj_BH<0.10 ~ "SigPadjBH10_Yes", .default = "SigPadjBH10_No"),
         SigPadjBH05 = case_when(Padj_BH<0.05 ~ "SigPadjBH05_Yes", .default = "SigPadjBH05_No")) %>%
  filter(grepl("T21", term)==TRUE) %>%
  group_by(chromosome, term, SigPadjBH10) %>%
  summarise(N_Aptamers_SigPadjBH10_T21byAge = n()) %>%
  ungroup() %>%
  spread(key = SigPadjBH10, value = N_Aptamers_SigPadjBH10_T21byAge) %>%
  mutate(N_Total_Aptamers_Chr = SigPadjBH10_No + SigPadjBH10_Yes,
         Pct_SigPadjBH.char = paste0(as.character(round(100*SigPadjBH10_Yes/N_Total_Aptamers_Chr)),
                                "%"),
         Pct_SigPadjBH = 100*SigPadjBH10_Yes/N_Total_Aptamers_Chr,
         gg_chromosome = as.numeric(gsub("Chr", "", chromosome)),
         gg_chromosome = case_when(is.na(gg_chromosome) ~ -1, .default = gg_chromosome)) %>%
  arrange(gg_chromosome) %>%
  mutate(chr21cat = case_when(chromosome == "Chr21" ~ "Chromosome 21",
                              .default = "Other chromosome"));

ggbar_modelB_T21sigPadjBH10_chrDist <- ggData_chr_distribution_SigPadjBH10.modelB02 %>%
  filter(term == "T21") %>%
  ggplot(aes(x = gg_chromosome, y = Pct_SigPadjBH, color = chr21cat, fill = chr21cat)) +
  geom_bar(position = "dodge", stat = "identity") +
  theme_bw() +
  theme(legend.position = "none",
        aspect.ratio = 1.0) +
  xlab("Chromosome") +
  ylab("Aptamers (%)") +
  ggtitle("Chromosomal distribution of aptamers with\nsignificantly differential abundance in T21") +
  scale_x_continuous("Chromosome", 
                       labels = as.character(seq(-1, 24, by=1)),
                       breaks = seq(-1, 24, by = 1), 
                       expand=c(0,0))

ggbar_modelB_T21byAgesigPadjBH10_chrDist <- ggData_chr_distribution_SigPadjBH10.modelB02 %>%
  filter(term == "T21:Age") %>%
  ggplot(aes(x = gg_chromosome, y = Pct_SigPadjBH, color = chr21cat, fill = chr21cat)) +
  geom_bar(position = "dodge", stat = "identity") +
  theme_bw() +
  theme(legend.position = "none",
        aspect.ratio = 1.0) +
  xlab("Chromosome") +
  ylab("Aptamers (%)") +
  ggtitle("Chromosomal distribution of aptamers with\nsignificantly different age trajectory in T21") +
  scale_x_continuous("Chromosome", 
                       labels = as.character(seq(-1, 24, by=1)),
                       breaks = seq(-1, 24, by = 1), 
                       expand=c(0,0))

ggbar_modelB_T21sigPadjBH10_chrDist
ggbar_modelB_T21byAgesigPadjBH10_chrDist

setwd("/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/output")
ggsave(ggbar_modelB_T21sigPadjBH10_chrDist,
         filename = "20250124_ggbar_modelB_T21_sigPadjBH10_chrDist_v0.2.png",
         width = 5, height = 5, units = "in")
setwd("/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/output")
ggsave(ggbar_modelB_T21byAgesigPadjBH10_chrDist,
         filename = "20250124_ggbar_modelB_T21byAge_sigPadjBH10_chrDist_v0.2.png",
         width = 5, height = 5, units = "in")
```

```{r}
tidy.Zlog10Abundance_modelB02 %>%
  filter(term == "T21:Age" & Sig10_PadjBH == 1) %>%
  arrange(P) %>%
  select(-c(std.error, statistic)) %>%
  select(Aptamer_Chr_Cluster, everything()) %>%
  mutate(tmp = gsub("[(]", "_", Aptamer_Chr_Cluster)) %>%
  select(tmp, everything())

ar1_p4c_soma %>% select(Aptamer, UniProt)
```


```{r}
# https://pmc.ncbi.nlm.nih.gov/articles/PMC7062043/#S7
  
# "To analyze each incremental list of proteins, we used the R topGO package49 for GO analysis and the R clusterProfiler package50 for KEGG and Reactome analyses."

# topGO
# clusterProfiler

BiocManager::install("clusterProfiler")

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
#BiocManager::install("topGO")
#BiocManager::install("clusterProfiler")

#source("https://bioconductor.org/biocLite.R")
biocLite("BiocUpgrade") ## you may need this
biocLite("clusterProfiler")

library(topGO)
library(clusterProfiler)
browseVignettes("topGO")

# https://www.nature.com/articles/s41596-024-01020-z

# https://biostatsquid.com/pathway-enrichment-analysis-tutorial-clusterprofiler/#step1
# Split the dataframe into a list of sub-dataframes: upregulated, downregulated genes
deg_results_list <- split(df, df$diffexpressed)

## Run ClusterProfiler -----------------------------------------------
# Settings
name_of_comparison <- 'severevshealthy' # for our filename
background_genes <- 'reactome' # for our filename
bg_genes <- readRDS(paste0(bg_path, 'reactome.RDS')) # read in the background genes
padj_cutoff <- 0.05 # p-adjusted threshold, used to filter out pathways
genecount_cutoff <- 5 # minimum number of genes in the pathway, used to filter out pathways
filename <- paste0(out_path, 'clusterProfiler/', name_of_comparison, '_', background_genes) # filename of our PEA results
```


```{r}
# https://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/119-pca-in-r-using-ade4-quick-scripts/

#install.packages("ade4")      # PCA computation
#install.packages("factoextra")# PCA visualization
#install.packages("magrittr")  # for piping %>%
library(ade4)
library(factoextra)
library(magrittr)

library("factoextra")
data(decathlon2)
decathlon2.active <- decathlon2[1:23, 1:10]
head(decathlon2.active[, 1:6])
```

#### Prepare data for PCA of *T21s only*
```{r}
ar_p4c_soma_t21_pca <- ar1_p4c_soma %>%
  filter(Karyotype == "T21") %>%
  select(LabID, Aptamer, Z_log10_Abundance) %>%
  spread(key = Aptamer, value = Z_log10_Abundance) %>%
  unique() %>%
  column_to_rownames("LabID")

ar_p4c_soma_t21_pca
```

```{r}
# https://pmc.ncbi.nlm.nih.gov/articles/PMC7062043/#S7
# "To identify specific genetic variants associated with the aging plasma proteome, we mined the summary statistics generated by Sun et al.27, who found 1,927 associations with 1,104 plasma proteins. Qgraph59 R package was used to create a network between the genome and the 2,925 proteins analyzed in this study."

# To-Do: Request access
# https://ega-archive.org/studies/EGAS00001002555


```


```{r}
library(ade4)
res.pca <- dudi.pca(ar_p4c_soma_t21_pca,
                    scannf = FALSE,   # Hide scree plot
                    nf = 10            # Number of components kept in the results
                    )

# Visualize using factoextra
fviz_eig(res.pca)

# Graph of individuals. Individuals with a similar profile are grouped together.
fviz_pca_ind(res.pca,
             col.ind = "cos2", # Color by the quality of representation
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
             )

# Graph of variables. Positive correlated variables point to the same side of the plot. Negative correlated variables point to opposite sides of the graph.
fviz_pca_var(res.pca,
             col.var = "contrib", # Color by contributions to the PC
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
             )

# Biplot of individuals and variables
fviz_pca_biplot(res.pca, repel = TRUE,
                col.var = "#2E9FDF", # Variables color
                col.ind = "#696969"  # Individuals color
                )

# Access to the PCA results
library(factoextra)
# Eigenvalues
eig.val <- get_eigenvalue(res.pca)
eig.val
  
# Results for Variables
res.var <- get_pca_var(res.pca)
res.var$coord          # Coordinates
res.var$contrib        # Contributions to the PCs
res.var$cos2           # Quality of representation 
# Results for individuals
res.ind <- get_pca_ind(res.pca)
res.ind$coord          # Coordinates
res.ind$contrib        # Contributions to the PCs
res.ind$cos2           # Quality of representation 
```



```{r}
# https://pmc.ncbi.nlm.nih.gov/articles/PMC7062043/#S7
  
# "Type II sum of squares (SS) were calculated using the Anova function of the R car package44. This SS type tests for each main effect after the other main effects. Q-values were estimated using Benjamini–Hochberg approach45."
```

```{r}
# " To determine the relative proportion of variance explained by age and sex, we calculated the partial Eta2 as follows:"

# PartialEta2 = SSeffect / (SSeffect + SSerror)
 

```


#### FOR LIMITATIONS SECTION
```{r}
# https://www.researchsquare.com/article/rs-4862220/v1
```

```{r}
tidy.Zlog10Abundance_modelB %>%
  filter(term == "T21") %>%
  arrange(p.value) %>%
  select(-c(std.error, statistic)) %>%
  select(Aptamer_Cluster, everything()) %>%
  mutate(Sig10_PadjBH = case_when(Padj_BH<0.10 ~ 1, .default = 0),
         Sig05_PadjBH = case_when(Padj_BH<0.05 ~ 1, .default = 0)) %>%
  filter(Sig10_PadjBH == 1);

tidy.Zlog10Abundance_modelB %>%
  filter(term == "T21:Age") %>%
  arrange(p.value) %>%
  select(-c(std.error, statistic)) %>%
  select(Aptamer_Cluster, everything()) %>%
  mutate(Sig10_PadjBH = case_when(Padj_BH<0.10 ~ 1, .default = 0),
         Sig05_PadjBH = case_when(Padj_BH<0.05 ~ 1, .default = 0)) %>%
  filter(Sig10_PadjBH == 1) %>%
  mutate(rank = abs(estimate)*(-log10(Padj_BH))) %>%
  arrange(desc(rank)); 
```

```{r}
set.seed(1234)
fit.sumZcluster_modelB <-  ar_p4c_soma_sumZcluster %>%
  split(., .$Cluster) %>%
    map(~lm(sumZ_Cluster ~ T21 + Age + Female + T21*Age, data = .));

tidy.sumZcluster_modelB <-  ar_p4c_soma_sumZcluster %>%
  split(., .$Cluster) %>%
  map2(.x = fit.sumZcluster_modelB, .y = .,
       .f = ~tidy(x = .x, conf.int=TRUE, data = .y)) %>%
    bind_rows(.id="Cluster") %>%
    filter(term!="(Intercept)") %>%
  group_by(term) %>%
  mutate(Padj_BH = p.adjust(p.value, method = "BH")) %>%
  ungroup();

tidy.sumZcluster_modelB %>%
  filter(term == "T21") %>%
  arrange(p.value) %>%
  select(-c(std.error, statistic));
tidy.sumZcluster_modelB %>%
  filter(term == "T21:Age") %>%
  arrange(p.value) %>%
  select(-c(std.error, statistic));
```

```{r}
set.seed(1234)
fit.modelB <-  ar1_p4c_soma %>%
  split(., .$Cluster) %>%
    map(~lm(sumZ_Cluster ~ T21 + Age + Female + T21*Age, data = .));

tidy.modelB <-  ar_p4c_soma_sumZcluster %>%
  split(., .$Cluster) %>%
  map2(.x = fit.modelB, .y = .,
       .f = ~tidy(x = .x, conf.int=TRUE, data = .y)) %>%
    bind_rows(.id="Cluster") %>%
    filter(term!="(Intercept)") %>%
  group_by(term) %>%
  mutate(Padj_BH = p.adjust(p.value, method = "BH")) %>%
  ungroup();

tidy.modelB %>%
  filter(term == "T21") %>%
  arrange(p.value) %>%
  select(-c(std.error, statistic));
tidy.modelB %>%
  filter(term == "T21:Age") %>%
  arrange(p.value) %>%
  select(-c(std.error, statistic));
```



#### LASSO regression to classify T21 +/- Celiac
(See http://www.sthda.com/english/articles/36-classification-methods-essentials/149-penalized-logistic-regression-essentials-in-r-ridge-lasso-and-elastic-net/)

#### Perform stratified sampling
```{r}
# ar1_p4c_soma

training_IDs <- list()
test_IDs <- list()

set.seed(1234)

update_seq <- c(1, 10, seq(from=0, to=1000, by=10)[-c(1:2)] )
update_seq

ar1_p4c_soma_T21 <- ar1_p4c_soma %>%
  filter(T21 == 1) %>%
  mutate(rownum = row_number()) %>%
  select(rownum, everything())

ar1_p4c_soma_T21_IDs <- ar1_p4c_soma_T21 %>% select(LabID) %>% unique()

for ( i in 1:1000 ){
#for ( i in 1:3 ){
  seedvalue = 1234*i
  set.seed(seedvalue)
  # Split the data into training and test sets
  training_IDs.tmp <- ar1_p4c_soma_T21 %>%
    sample_frac(0.8) %>%
    mutate(train_test = "Train") %>%
    select(LabID, train_test, everything())
  test_IDs.tmp <- ar1_p4c_soma_T21 %>%
    filter(rownum %notin% training_data.tmp$rownum) %>%
    mutate(train_test = "Test") %>%
    select(LabID, train_test, everything())
  
  training_IDs[[i]] <- training_IDs.tmp
  test_IDs[[i]] <- test_IDs.tmp

  if ( i %in% update_seq ) {
    print(paste(Sys.time(), "  iter = ", i, " complete.", sep=""))
  }
}

training_data %>% length()
test_data %>% length()
```








```{r}
ar1_p4c_soma_chr <- ar1_p4c_soma %>%
  mutate(In_ar1_p4c_soma = 1) %>%
  left_join(gene_chr_anno, by = "UniProt")

ar1_p4c_soma_chr

ar1_p4c_soma_chr %>%
  mutate(chromosome21 = case_when(chromosome == 21 ~ "Chromosome 21",
                                  .default = "Other chromosome")) %>%
  select(Cluster, UniProt, chromosome, chromosome21, Z_log10_Abundance) %>%
  group_by(Cluster, chromosome21) %>%
  summarise(N_aptamers = n()) %>%
  ungroup() %>%
  spread(key = chromosome21, value = N_aptamers)

ar1_p4c_soma_chr %>%
  mutate(chromosome21 = case_when(chromosome == 21 ~ "Chromosome 21",
                                  .default = "Other chromosome")) %>%
  select(Cluster, UniProt, chromosome, chromosome21, Z_log10_Abundance) %>%
  group_by(Cluster, chromosome) %>%
  summarise(N_aptamers = n()) %>%
  ungroup() %>%
  group_by(Cluster) %>%
  arrange(Cluster, chromosome) %>%
  mutate(N_total_cluster_aptamers = sum(N_aptamers, na.rm = TRUE)) %>%
  ungroup() %>%
  mutate(Percent_cluster_aptamers = round(100*N_aptamers/N_total_cluster_aptamers, digits = 1)) %>%
  mutate(gg_chromosome = paste0("chr", chromosome),
         gg_chromosome = factor(gg_chromosome, levels = c(paste0("chr", seq(1:24))))) %>%
  arrange(gg_chromosome) %>%
  ggplot(aes(x = gg_chromosome, y = Percent_cluster_aptamers, fill = Cluster)) +
  geom_bar(position = "dodge", stat = "identity") +
  facet_wrap(~Cluster, ncol = 4) +
  coord_flip()

ar1_p4c_soma_chr %>%
  mutate(chromosome21 = case_when(chromosome == 21 ~ "Chromosome 21",
                                  .default = "Other chromosome")) %>%
  select(Cluster, UniProt, chromosome, chromosome21, Z_log10_Abundance) %>%
  group_by(Cluster, chromosome) %>%
  summarise(N_aptamers = n()) %>%
  ungroup() %>%
  group_by(Cluster) %>%
  arrange(Cluster, chromosome) %>%
  mutate(N_total_cluster_aptamers = sum(N_aptamers, na.rm = TRUE)) %>%
  ungroup() %>%
  mutate(Percent_cluster_aptamers = round(100*N_aptamers/N_total_cluster_aptamers, digits = 1)) %>%
  mutate(gg_chromosome = paste0("chr", chromosome),
         gg_chromosome = factor(gg_chromosome, levels = c(paste0("chr", seq(1:24))))) %>%
  arrange(gg_chromosome) %>%
  ggplot(aes(x = Cluster, y = Percent_cluster_aptamers, fill = gg_chromosome)) +
  geom_bar(position = "dodge", stat = "identity") +
  facet_wrap(~gg_chromosome, ncol = 4) +
  coord_flip()

ar1_p4c_soma_chr %>%
  mutate(chromosome21 = case_when(chromosome == 21 ~ "Chromosome 21",
                                  .default = "Other chromosome")) %>%
  select(Cluster, UniProt, chromosome, chromosome21, Z_log10_Abundance) %>%
  group_by(Cluster, chromosome) %>%
  summarise(N_aptamers = n()) %>%
  ungroup() %>%
  group_by(Cluster) %>%
  arrange(Cluster, chromosome) %>%
  mutate(N_total_cluster_aptamers = sum(N_aptamers, na.rm = TRUE)) %>%
  ungroup() %>%
  mutate(Percent_cluster_aptamers = round(100*N_aptamers/N_total_cluster_aptamers, digits = 1)) %>%
  mutate(gg_chromosome = paste0("chr", chromosome),
         gg_chromosome = factor(gg_chromosome, levels = c(paste0("chr", seq(1:24))))) %>%
  arrange(gg_chromosome) %>%
  filter(gg_chromosome == "chr21") %>%
  ggplot(aes(x = Cluster, y = Percent_cluster_aptamers, fill = gg_chromosome)) +
  geom_bar(position = "dodge", stat = "identity") +
  facet_wrap(~gg_chromosome, ncol = 4) +
  coord_flip()
```

```{r}
ar1_p4c_soma_chr
```









```{r}
tmp <- ar1_p4c_soma %>%
  filter(grepl("nterferon", TargetFullName)==TRUE |
           grepl("reactive protein", TargetFullName) == TRUE)
tmp

set.seed(1234)
fit.tmp <- tmp %>%
  split(., .$Aptamer) %>%
  map(~lm(log10_Abundance ~ T21 + Age + Female, data = .))
tidy.tmp <-  tmp %>%
  split(., .$Aptamer) %>%
  map2(.x = fit.tmp, .y = .,
       .f = ~tidy(x = .x, conf.int=TRUE, data = .y)) %>%
    bind_rows(.id="Aptamer") %>%
    filter(term!="(Intercept)") %>%
  mutate(Padj_BH = p.adjust(p.value, method = "BH")) %>%
  ungroup();

tidy.tmp %>%
  filter(term == "T21") %>%
  arrange(p.value)

ar1_p4c_soma %>%
  filter(grepl("reactive protein", TargetFullName)==TRUE)

```


```{r}
donovan2024_p4c_meta
```

```{r}
select <- dplyr::select

p4c_soma_genes <- p4c_soma %>%
  select(GeneSymbol) %>%
  unique()
p4c_soma_genes

fwrite(p4c_soma_genes, "/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/output/20250124_published_p4c_soma_genes.csv")
fwrite(p4c_soma_genes, "/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/output/20250124_published_p4c_soma_genes.txt",
       sep = "\t")

p4c_soma_genes_chr <- fread("/Users/jessica/Dropbox/EspinosaGroup/ANALYSIS/2025/dev/rawdata/martquery_0124230151_605.txt.gz") %>%
  select(`Gene name`, `Chromosome/scaffold name`) %>%
  unique() %>%
  mutate(In_biomart_export = 1) %>%
  full_join(p4c_soma_genes %>% mutate(In_p4c_soma_genes = 1), by = c("Gene name" = "GeneSymbol")) %>%
  filter(!is.na(In_p4c_soma_genes)) %>%
  mutate(chromosome = gsub("CHR", "|CHR", `Chromosome/scaffold name`)) %>%
  rename(GeneSymbol = `Gene name`) %>%
  mutate(chromosome.clean = case_when(grepl("CHR11_", chromosome)==TRUE | grepl("CHR11", chromosome)==TRUE ~ 11,
                                      grepl("CHR11_", chromosome)==TRUE | grepl("CHR11", chromosome)==TRUE ~ 11,
                                      grepl("CHR11_", chromosome)==TRUE | grepl("CHR11", chromosome)==TRUE ~ 11,
                                      grepl("CHR11_", chromosome)==TRUE | grepl("CHR11", chromosome)==TRUE ~ 11,
                                      grepl("CHR11_", chromosome)==TRUE | grepl("CHR11", chromosome)==TRUE ~ 11,
                                      grepl("CHR11_", chromosome)==TRUE | grepl("CHR11", chromosome)==TRUE ~ 11,
                                      grepl("CHR17_", chromosome)==TRUE | grepl("CHR11", chromosome)==TRUE ~ 11,
                                      grepl("CHR18_", chromosome)==TRUE | grepl("CHR11", chromosome)==TRUE ~ 11,
                                      grepl("CHR19_", chromosome)==TRUE | grepl("CHR11", chromosome)==TRUE ~ 11,
                                      
                                      grepl("CHR1_", chromosome)==TRUE ~ 1,
                                      chromosome == "Y" ~ 23,
                                      chromosome == "X" ~ 22,
                                      .default = NA))
p4c_soma_genes_chr

p4c_soma_genes_chr %>% select(chromosome, chromosome.clean) %>% unique() %>% arrange(chromosome.clean)

p4c_soma_genes_chr %>%
  group_by(In_biomart_export, In_p4c_soma_genes) %>%
  summarise(N = n())
```


```{r}
ar1_p4c_soma_genes <- (ar1_p4c_soma %>%
  dplyr::select(GeneSymbol) %>%
  unique())$GeneSymbol

ar1_p4c_soma_genes_cluster1 <- (ar1_p4c_soma %>%
                         filter(Cluster == "Cluster 1") %>%
                         dplyr::select(GeneSymbol) %>%
                         unique())$GeneSymbol

# install the biomaRt package       
# source("https://bioconductor.org/biocLite.R")     
# biocLite("biomaRt")       
# load biomaRt      
library(biomaRt)        

# 1) select a mart and data set        
 mart <- biomaRt::useDataset(dataset = "hsapiens_gene_ensembl",         
                    mart    = useMart("ENSEMBL_MART_ENSEMBL",       
                    host    = "https://www.ensembl.org"))       
        
 head(biomaRt::listAttributes(biomaRt::useDataset(
                                         dataset = "hsapiens_gene_ensembl",         
                                         mart    = useMart("ENSEMBL_MART_ENSEMBL",      
                                         host    = "https://www.ensembl.org"))), 10)
 
# 2) run a biomart query using the getBM() function        
# and specify the attributes and filter arguments      
cluster1_anno <- list() 
for ( i in 1:length(ar1_p4c_soma_genes_cluster1) ) {
  cluster1_anno[[i]] <- biomaRt::getBM(attributes = c("chromosome_name", "description"),       
                      filters    = "hgnc_symbol",       
                      values     = ar1_p4c_soma_genes_cluster1[[i]],         
                      mart       = mart)
}
```



```{r}
ggData %>%
  filter(Cluster == "Cluster 8") %>%
  select(TargetFullName) %>%
  unique()

# https://www.mdpi.com/1422-0067/19/1/248
# Wyganowska-Świątkowska, M., Matthews-Kozanecka, M., Matthews-Brzozowska, T., Skrzypczak-Jankun, E., & Jankun, J. (2018). Can egcg alleviate symptoms of down syndrome by altering proteolytic activity? International Journal of Molecular Sciences, 19(1), 248. https://doi.org/10.3390/ijms19010248
```


